### PR TITLE
Prevent "Continue setup" navigating back immediately

### PR DESCRIPTION
**Pull request by neil-marcellini**
### Details

Prevent immediately navigating back to OldDot when signed out and following the "Continue setup" transition flow.

### Related Issue

$ https://github.com/Expensify/App/issues/8295

### Tests

First, set up your dev environment so that the betas load slowly.

1. Set `$allBetasEnabled` to false [here](https://github.com/Expensify/Web-Expensify/blob/31444587b13c5ec4ed0a3ebbce3b7ed828b9631a/lib/BetaManager.php#L96)
2. Add these betas `[BetaManager::BETA_FREE_PLAN, BetaManager::BETA_FREE_PLAN_FULL_LAUNCH];` to the Web-Expensify BetaManager [here](https://github.com/Expensify/Web-Expensify/blob/31444587b13c5ec4ed0a3ebbce3b7ed828b9631a/lib/BetaManager.php#L102-L103)
3. Prevent all betas from automatically being enabled on dev in NewDot [here](https://github.com/Expensify/App/blob/a671ac659f49157c6106c9ba1abbc7cbaf6cb975/src/libs/Permissions.js#L11)
4. Add a 5 second delay to loading the betas by replacing [this line](https://github.com/Expensify/App/blob/fb5177e13534e57096ed666f709513c8261b7f2c/src/libs/actions/User.js#L92) with the following:

```
setTimeout(() => Onyx.set(ONYXKEYS.BETAS, response.betas), 5 * 1000);
```

Now test:
1. Make sure you are signed out of NewDot
2. Go to OldDot and sign up for a new account with an @gmail.com address
3. Click on the "Get started" on the inbox task that says "Would you like to get started with our free plan?".
5. Sign out of NewDot
6. Go back to the OldDot inbox and refresh the page
7. Click "Continue setup" on the inbox task that says "Finish setting up your bank account".
8. Verify that the connect bank account modal is displayed after 5 seconds of loading, and you are not immediately navigated back to OldDot.

- [ ] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [ ] I linked the correct issue in the `### Fixed Issues` section above
- [ ] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps

- [ ] Verify that no errors appear in the JS console

### CODE REVIEW

**Review by TomatoToaster — APPROVED** [review]
Oh my b, I thought I already gave this an approve, LGTM!

**Comment by TomatoToaster** [comment]
@neil-marcellini are we going to have [No QA] for this? I wasn't sure if you wanted to test the part after `setTimeout(() => Onyx.set(ONYXKEYS.BETAS, response.betas), 5 * 1000);` on staging or not. If not, we should mark it as `[No Qa]` in the title.

**Comment by neil-marcellini** [comment]
@TomatoToaster since this PR only merges into the feature branch `neil-transition-flows` there won't be any QA right now. When I put up a PR for that feature branch I will include QA steps for all of the transition flows, and I'll probably leave out the step to make the betas load slowly since that can't be QAed. Thanks for the review!

File: src/pages/LogInWithShortLivedTokenPage.js
Comment (marcaaron) [inline src/pages/LogInWithShortLivedTokenPage.js:134]: Oh the betas are back now? 😂

File: src/pages/LogInWithShortLivedTokenPage.js
Comment (marcaaron) [inline src/pages/LogInWithShortLivedTokenPage.js:89]: **thought:** This would be easier to understand if we do not use `componentDidUpdate()` to mean "wait to navigate until the betas are loaded". It's not clear what specifically we are "waiting" for or where it happens. Who calls `getBetas()` and why are we waiting for that to happen here? 

**suggestion:** The comment about `ReimbursementAccountPage` creates a non-code dependency that should be changed as it's impossible to keep up to date. If `ReimbursementAccountPage` some day has no need for the betas will anyone remember to update this comment?

File: src/pages/LogInWithShortLivedTokenPage.js
Comment (marcaaron) [inline src/pages/LogInWithShortLivedTokenPage.js:85]: Other things could later update this component... should we "navigate to the exit route" if they do?

File: src/pages/LogInWithShortLivedTokenPage.js
Comment (neil-marcellini) [reply src/pages/LogInWithShortLivedTokenPage.js:134]: 😅 yes unfortunately

File: src/pages/LogInWithShortLivedTokenPage.js
Comment (neil-marcellini) [reply src/pages/LogInWithShortLivedTokenPage.js:89]: You are right that this is very fragile and doesn't make much sense. 

What do we do if someone who isn't on the betas required for the ReimbursementAccountPage tries to navigate there? Currently we kick them back to the home page immediately, but that is problematic if they will have access to that page once the betas load. We could instead show a loading screen on the ReimbursementAccountPage until the betas have loaded. Is that a good solution?

File: src/pages/LogInWithShortLivedTokenPage.js
Comment (neil-marcellini) [reply src/pages/LogInWithShortLivedTokenPage.js:85]: The only things that can make this component update are the props correct? So as long as we don't add more props / Onyx keys it will be fine.

What if we moved all of this into an action called "signInTransitioningUser" that would handle signing in, out, and navigating? That way instead of using componentDidUpdate to wait for changes we could use promises / Onyx callbacks to call these actions in sequence. For example,  sign out old user > wait for onyx to clear > sign in new user > navigate. I'm not sure if that's any different than what we are doing here, but please let me know what you think.

**Comment by neil-marcellini** [comment]
Thanks for the comments @marcaaron. Sorry that I forgot to wait before merging.

File: src/pages/LogInWithShortLivedTokenPage.js
Comment (marcaaron) [reply src/pages/LogInWithShortLivedTokenPage.js:89]: Gonna think about this for a bit.

File: src/pages/LogInWithShortLivedTokenPage.js
Comment (marcaaron) [reply src/pages/LogInWithShortLivedTokenPage.js:85]: >The only things that can make this component update are the props correct

`componentDidUpdate()` runs after a component's `render()` method is called. a change in props or state will cause the component to render, but so can other things like a parent component re-rendering (with no change at all to props or state).

>So as long as we don't add more props / Onyx keys it will be fine.

That's a good way to rephrase my concern 😄  

How do we make everyone understand that the possibility exists to add more props and for things to not be fine?

>What if we moved all of this into an action called "signInTransitioningUser" that would handle signing in, out, and navigating? That way instead of using componentDidUpdate to wait for changes we could use promises / Onyx callbacks to call these actions in sequence. For example, sign out old user > wait for onyx to clear > sign in new user > navigate. I'm not sure if that's any different than what we are doing here, but please let me know what you think.

This sounds interesting and I think it is pretty different compared to what we are doing here. It also sounds much easier to follow because the code paths will be separated cleanly and we won't have to worry about which "case" we are handling as we bounce between different lifecycle methods. Let's chat more 1:1 about this.

**Pull request merged**
